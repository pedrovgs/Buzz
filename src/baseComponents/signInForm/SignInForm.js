import React from "react";
import { TextField, RaisedButton } from "material-ui";
import { User } from "../../session/actions";
import firebaseAuth from "../../firebase/firebaseAuth";

const initialState = {
  isLogInInProgress: false,
  email: undefined,
  password: undefined,
  errorMessage: undefined
};

const formStyle = {
  padding: "6px"
};

const buttonStyle = {
  margin: 12
};

class SignInForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = initialState;
    this.onEmailFieldChanged = this.onEmailFieldChanged.bind(this);
    this.onPasswordFieldChanged = this.onPasswordFieldChanged.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  render() {
    return (
      <form style={formStyle} id="signInForm">
        <div>
          <TextField
            id="emailField"
            floatingLabelText="Email"
            onChange={this.onEmailFieldChanged}
            errorText={this.state.errorMessage}
          />
        </div>
        <div>
          <TextField
            id="passwordField"
            type="password"
            floatingLabelText="Password"
            onChange={this.onPasswordFieldChanged}
          />
        </div>
        <div>
          <RaisedButton
            id="signUpButton"
            label="Sign Up"
            style={buttonStyle}
            onClick={this.onSubmit}
          />
          <RaisedButton
            id="signInButton"
            label="Sign In"
            style={buttonStyle}
            onClick={this.onSubmit}
          />
        </div>
      </form>
    );
  }

  onEmailFieldChanged(event, newValue) {
    this.setState(() => ({
      email: newValue
    }));
  }

  onPasswordFieldChanged(event, newValue) {
    this.setState(() => ({
      password: newValue
    }));
  }

  onSubmit(event) {
    if (event) {
      event.preventDefault();
    }
    this.setStateToLogInInProgress(true);
    this.signInOrSignUpIfTheUserAlreadyExists();
  }

  signInOrSignUpIfTheUserAlreadyExists() {
    const email = this.state.email;
    const pass = this.state.password;
    firebaseAuth.signOut().then(() => {
      firebaseAuth
        .signIn(email, pass)
        .then(() => {
          this.notifyLogInSuccess();
        })
        .catch(error => {
          let errorCode = error.code;
          if (errorCode === "auth/user-not-found") {
            this.createAccount(email, pass);
          } else if (errorCode) {
            this.setStateToError(error.message);
          }
          this.setStateToLogInInProgress(false);
        });
    });
  }

  createAccount(email, password) {
    firebaseAuth
      .signUp(email, password)
      .then(() => {
        this.notifyLogInSuccess();
      })
      .catch(error => {
        const errorMessage = error.message;
        if (errorMessage) {
          this.setStateToError(errorMessage);
        } else {
          this.notifyLogInSuccess();
        }
        this.setStateToLogInInProgress(false);
      });
  }

  notifyLogInSuccess() {
    const user = firebaseAuth.currentUser();
    if (user) {
      this.props.onUserLoggedIn(new User(user.email));
    }
  }

  setStateToLogInInProgress(logInInProgress) {
    this.setState(() => ({
      isLogInInProgress: logInInProgress
    }));
  }

  setStateToError(error) {
    this.setState(() => ({
      errorMessage: error
    }));
  }
}

export default SignInForm;
