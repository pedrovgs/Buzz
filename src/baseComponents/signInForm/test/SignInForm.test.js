import React from "react";
import Enzyme from "enzyme";
import { createSerializer } from "enzyme-to-json";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import SignInForm from "../SignInForm";
jest.mock("material-ui/TextField", () => "TextField");
jest.mock("material-ui/RaisedButton", () => "RaisedButton");

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));
Enzyme.configure({ adapter: new Adapter() });

describe("SignInForm", () => {
  it("shows an empty form with the email, pass, sign up and sing in buttons by default", () => {
    const signUpForm = renderSignupForm();

    expect(signUpForm).toMatchSnapshot();
  });

  function renderSignupForm() {
    return toJson(Enzyme.shallow(<SignInForm />));
  }
});
