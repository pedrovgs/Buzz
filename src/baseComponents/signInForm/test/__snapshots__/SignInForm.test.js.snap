// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm shows an empty form with the email, pass, sign up and sing in buttons by default 1`] = `
<form
  id="signInForm"
  style={
    Object {
      "padding": "6px",
    }
  }
>
  <div>
    <TextField
      floatingLabelText="Email"
      onChange={[Function]}
    />
  </div>
  <div>
    <TextField
      floatingLabelText="Password"
      onChange={[Function]}
      type="password"
    />
  </div>
  <div>
    <RaisedButton
      label="Sign Up"
      onClick={[Function]}
      style={
        Object {
          "margin": 12,
        }
      }
    />
    <RaisedButton
      label="Sign In"
      onClick={[Function]}
      style={
        Object {
          "margin": 12,
        }
      }
    />
  </div>
</form>
`;
